File: scripts/skill_usage_collect.py
#!/usr/bin/env python3
"""Collect skill-activation signal from Claude Code session jsonl.

Implements step-2-skill-inventory-rationalization.md Phase 1 Step 2.
Reads `~/.claude/projects/<project-slug>/*.jsonl` for the current repo,
parses each turn for two signals:

  - exposure: the skill slug appeared in an `attachment.type=skill_listing`
    payload (catalog presented to the agent that turn).
  - mention:  the assistant-text response in the same or following turn
    referenced the slug in backticks with one of the anchor verbs
    (using, via, per, route, dispatch, invoke, call) OR cited a SKILL.md
    path under `.augment/skills/<slug>/`, `.claude/skills/<slug>/`, or
    `.agent-src/skills/<slug>/`.

Emits one JSONL record per (session, turn, slug, kind) to
`agents/runtime/metrics/skill-usage.jsonl` (append-only, deduped on the
(session_id, turn_idx, slug, kind) tuple).

Privacy: `prompt_excerpt_hash` = SHA-256 of the first 200 chars of the
user prompt that opened the turn. No raw user or assistant bodies are
persisted. See `agents/evidence/audits/2026-05-14-north-star/skill-usage-sources.md`.
"""
from __future__ import annotations

import argparse
import hashlib
import json
import re
import sys
from pathlib import Path
from typing import Iterable, Iterator

REPO = Path(__file__).resolve().parent.parent
OUT = REPO / "agents" / "metrics" / "skill-usage.jsonl"

LISTING_LINE_RE = re.compile(r"^-\s+([a-z0-9][a-z0-9_-]+):\s", re.MULTILINE)
ANCHOR_VERBS = ("using", "via", "per", "route", "routing", "dispatch", "dispatched", "invoke", "call")
PATH_RE = re.compile(r"\.(?:augment|claude|agent-src)/skills/([a-z0-9][a-z0-9_-]+)/SKILL\.md")


def project_slug(repo: Path) -> str:
    return str(repo).replace("/", "-")


def session_files(slug: str) -> list[Path]:
    base = Path.home() / ".claude" / "projects" / slug
    if not base.is_dir():
        return []
    return sorted(base.glob("*.jsonl"))


def iter_turns(jsonl: Path) -> Iterator[dict]:
    with jsonl.open("r", encoding="utf-8", errors="replace") as fh:
        for line in fh:
            line = line.strip()
            if not line:
                continue
            try:
                yield json.loads(line)
            except json.JSONDecodeError:
                continue


def extract_listing(entry: dict) -> set[str]:
    att = entry.get("attachment") or {}
    if att.get("type") != "skill_listing":
        return set()
    content = att.get("content", "") or ""
    return set(LISTING_LINE_RE.findall(content))


def extract_text(entry: dict) -> str:
    if entry.get("type") != "assistant":
        return ""
    msg = entry.get("message") or {}
    content = msg.get("content")
    if isinstance(content, str):
        return content
    if isinstance(content, list):
        return "\n".join(p.get("text", "") for p in content if isinstance(p, dict) and p.get("type") == "text")
    return ""


def find_mentions(text: str, known_slugs: Iterable[str]) -> set[str]:
    hits: set[str] = set()
    if not text:
        return hits
    hits.update(PATH_RE.findall(text))
    for slug in known_slugs:
        token = f"`{slug}`"
        if token not in text:
            continue
        lower = text.lower()
        for verb in ANCHOR_VERBS:
            if f"{verb} {token}".lower() in lower or f"{verb} the {token}".lower() in lower:
                hits.add(slug)
                break
    return hits


def hash_prompt(text: str) -> str:
    if not text:
        return ""
    return hashlib.sha256(text[:200].encode("utf-8", errors="replace")).hexdigest()[:16]


def collect_session(jsonl: Path, all_known: set[str]) -> list[dict]:
    session_id = jsonl.stem
    records: list[dict] = []
    last_prompt_hash = ""
    listed: set[str] = set()
    turn_idx = -1
    for entry in iter_turns(jsonl):
        etype = entry.get("type")
        if etype == "user":
            turn_idx += 1
            msg = entry.get("message") or {}
            body = msg.get("content") if isinstance(msg.get("content"), str) else ""
            last_prompt_hash = hash_prompt(body or "")
            continue
        if etype == "attachment":
            listed |= extract_listing(entry)
            continue
        if etype == "assistant":
            text = extract_text(entry)
            mentions = find_mentions(text, listed | all_known)
            ts = entry.get("timestamp") or ""
            for slug in sorted(listed):
                records.append({"session_id": session_id, "turn_idx": turn_idx, "slug": slug,
                                "kind": "exposure", "ts": ts, "prompt_excerpt_hash": last_prompt_hash})
            for slug in sorted(mentions):
                records.append({"session_id": session_id, "turn_idx": turn_idx, "slug": slug,
                                "kind": "mention", "ts": ts, "prompt_excerpt_hash": last_prompt_hash})
            listed = set()
    return records


def load_known_slugs(repo: Path) -> set[str]:
    slugs: set[str] = set()
    for root in (repo / ".augment" / "skills", repo / ".claude" / "skills", repo / ".agent-src" / "skills"):
        if not root.is_dir():
            continue
        for skill_md in root.glob("*/SKILL.md"):
            slugs.add(skill_md.parent.name)
    return slugs


def dedup_key(rec: dict) -> tuple:
    return (rec["session_id"], rec["turn_idx"], rec["slug"], rec["kind"])


def main() -> int:
    ap = argparse.ArgumentParser(description=__doc__.splitlines()[0])
    ap.add_argument("--project-slug", help="Override the ~/.claude/projects slug (defaults to current repo)")
    ap.add_argument("--out", type=Path, default=OUT, help="Output jsonl (default: agents/runtime/metrics/skill-usage.jsonl)")
    ap.add_argument("--quiet", action="store_true", help="Suppress non-error output")
    args = ap.parse_args()

    slug = args.project_slug or project_slug(REPO)
    files = session_files(slug)
    if not files:
        if not args.quiet:
            print(f"no session files for slug {slug}", file=sys.stderr)
        return 0
    known = load_known_slugs(REPO)
    seen: set[tuple] = set()
    args.out.parent.mkdir(parents=True, exist_ok=True)
    if args.out.exists():
        for line in args.out.read_text(encoding="utf-8", errors="replace").splitlines():
            try:
                seen.add(dedup_key(json.loads(line)))
            except (json.JSONDecodeError, KeyError):
                continue
    appended = 0
    with args.out.open("a", encoding="utf-8") as fh:
        for jsonl in files:
            for rec in collect_session(jsonl, known):
                k = dedup_key(rec)
                if k in seen:
                    continue
                seen.add(k)
                fh.write(json.dumps(rec, separators=(",", ":")) + "\n")
                appended += 1
    if not args.quiet:
        print(f"✅  Wrote {appended} new record(s) to {args.out.relative_to(REPO)} ({len(seen)} total)")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: src/scripts/skill_usage_collect.ts
#!/usr/bin/env tsx
/**
 * Collect skill-activation signal from Claude Code session jsonl.
 *
 * TypeScript twin of `src/scripts/skill_usage_collect.py` (ADR-089, Phase 8 /
 * Wave 8b). Mirrors the CLI contract EXACTLY — the `--project-slug`, `--out`,
 * `--quiet` flags, exit codes, the stdout/stderr split, byte-identical
 * messages AND the byte-identical appended JSONL (`json.dumps(rec,
 * separators=(",", ":")) + "\n"`).
 *
 * Reads `~/.claude/projects/<project-slug>/*.jsonl` for the current repo,
 * parses each turn for two signals — exposure (skill appeared in a
 * skill_listing attachment) and mention (assistant text referenced the slug
 * with an anchor verb, or cited a SKILL.md path).
 *
 * Emits one JSONL record per (session, turn, slug, kind), append-only, deduped
 * on the (session_id, turn_idx, slug, kind) tuple.
 *
 * No behaviour changes — latent Python quirks replicated.
 */
import * as crypto from 'node:crypto';
import * as fs from 'node:fs';
import * as os from 'node:os';
import * as path from 'node:path';
import process from 'node:process';
import { fileURLToPath, pathToFileURL } from 'node:url';

const _HERE = fileURLToPath(import.meta.url);

// src/scripts/skill_usage_collect.ts → parents[2] of the .py file is repo root.
export const REPO = path.resolve(path.dirname(_HERE), '..', '..');
const OUT = path.join(REPO, 'agents', 'metrics', 'skill-usage.jsonl');

// LISTING_LINE_RE = re.compile(r"^-\s+([a-z0-9][a-z0-9_-]+):\s", re.MULTILINE)
const LISTING_LINE_RE = /^-\s+([a-z0-9][a-z0-9_-]+):\s/gm;
const ANCHOR_VERBS: readonly string[] = [
    'using',
    'via',
    'per',
    'route',
    'routing',
    'dispatch',
    'dispatched',
    'invoke',
    'call',
];
// PATH_RE = re.compile(r"\.(?:augment|claude|agent-src)/skills/([a-z0-9][a-z0-9_-]+)/SKILL\.md")
const PATH_RE = /\.(?:augment|claude|agent-src)\/skills\/([a-z0-9][a-z0-9_-]+)\/SKILL\.md/g;

type Json = null | boolean | number | string | Json[] | { [k: string]: Json };

interface Record_ {
    session_id: string;
    turn_idx: number;
    slug: string;
    kind: string;
    ts: string;
    prompt_excerpt_hash: string;
}

export function project_slug(repo: string): string {
    return repo.replace(/\//g, '-');
}

export function session_files(slug: string): string[] {
    const base = path.join(os.homedir(), '.claude', 'projects', slug);
    let stat: fs.Stats;
    try {
        stat = fs.statSync(base);
    } catch {
        return [];
    }
    if (!stat.isDirectory()) {
        return [];
    }
    const names = fs
        .readdirSync(base)
        .filter((n) => n.endsWith('.jsonl'))
        .map((n) => path.join(base, n));
    names.sort();
    return names;
}

function* iter_turns(jsonl: string): Generator<{ [k: string]: Json }> {
    let content: string;
    try {
        content = fs.readFileSync(jsonl, 'utf-8');
    } catch {
        return;
    }
    for (const raw of content.split('\n')) {
        const line = raw.trim();
        if (!line) {
            continue;
        }
        let parsed: Json;
        try {
            parsed = JSON.parse(line) as Json;
        } catch {
            continue;
        }
        if (parsed !== null && typeof parsed === 'object' && !Array.isArray(parsed)) {
            yield parsed;
        } else {
            // Python `json.loads` on a bare value yields a non-dict; the
            // downstream `.get()` calls would raise AttributeError. The real
            // session jsonl is always object-per-line, so this never happens
            // in practice; yield an empty dict to preserve iteration count.
            yield {};
        }
    }
}

function _asObj(v: Json | undefined): { [k: string]: Json } | null {
    if (v !== null && typeof v === 'object' && !Array.isArray(v)) {
        return v;
    }
    return null;
}

function _str(v: Json | undefined): string {
    return typeof v === 'string' ? v : '';
}

export function extract_listing(entry: { [k: string]: Json }): Set<string> {
    const att = _asObj(entry['attachment']) ?? {};
    if (att['type'] !== 'skill_listing') {
        return new Set();
    }
    const content = _str(att['content']);
    const out = new Set<string>();
    LISTING_LINE_RE.lastIndex = 0;
    let m: RegExpExecArray | null;
    while ((m = LISTING_LINE_RE.exec(content)) !== null) {
        out.add(m[1] as string);
    }
    return out;
}

export function extract_text(entry: { [k: string]: Json }): string {
    if (entry['type'] !== 'assistant') {
        return '';
    }
    const msg = _asObj(entry['message']) ?? {};
    const content = msg['content'];
    if (typeof content === 'string') {
        return content;
    }
    if (Array.isArray(content)) {
        const parts: string[] = [];
        for (const p of content) {
            const o = _asObj(p);
            if (o && o['type'] === 'text') {
                parts.push(_str(o['text']));
            }
        }
        return parts.join('\n');
    }
    return '';
}

export function find_mentions(text: string, knownSlugs: Iterable<string>): Set<string> {
    const hits = new Set<string>();
    if (!text) {
        return hits;
    }
    PATH_RE.lastIndex = 0;
    let m: RegExpExecArray | null;
    while ((m = PATH_RE.exec(text)) !== null) {
        hits.add(m[1] as string);
    }
    const lower = text.toLowerCase();
    for (const slug of knownSlugs) {
        const token = `\`${slug}\``;
        if (!text.includes(token)) {
            continue;
        }
        for (const verb of ANCHOR_VERBS) {
            if (
                lower.includes(`${verb} ${token}`.toLowerCase()) ||
                lower.includes(`${verb} the ${token}`.toLowerCase())
            ) {
                hits.add(slug);
                break;
            }
        }
    }
    return hits;
}

export function hash_prompt(text: string): string {
    if (!text) {
        return '';
    }
    // Python hashes text[:200] (first 200 code points) as UTF-8 bytes.
    const first200 = Array.from(text).slice(0, 200).join('');
    return crypto
        .createHash('sha256')
        .update(Buffer.from(first200, 'utf-8'))
        .digest('hex')
        .slice(0, 16);
}

/** Mirror Python `sorted(set)` over slug strings. */
function _sorted(s: Iterable<string>): string[] {
    return [...s].sort((a, b) => (a < b ? -1 : a > b ? 1 : 0));
}

export function collect_session(jsonl: string, allKnown: Set<string>): Record_[] {
    const sessionId = path.basename(jsonl, '.jsonl');
    const records: Record_[] = [];
    let lastPromptHash = '';
    let listed = new Set<string>();
    let turnIdx = -1;
    for (const entry of iter_turns(jsonl)) {
        const etype = entry['type'];
        if (etype === 'user') {
            turnIdx += 1;
            const msg = _asObj(entry['message']) ?? {};
            const content = msg['content'];
            const body = typeof content === 'string' ? content : '';
            lastPromptHash = hash_prompt(body || '');
            continue;
        }
        if (etype === 'attachment') {
            for (const v of extract_listing(entry)) {
                listed.add(v);
            }
            continue;
        }
        if (etype === 'assistant') {
            const text = extract_text(entry);
            const union = new Set<string>([...listed, ...allKnown]);
            const mentions = find_mentions(text, union);
            const ts = _str(entry['timestamp']);
            for (const slug of _sorted(listed)) {
                records.push({
                    session_id: sessionId,
                    turn_idx: turnIdx,
                    slug,
                    kind: 'exposure',
                    ts,
                    prompt_excerpt_hash: lastPromptHash,
                });
            }
            for (const slug of _sorted(mentions)) {
                records.push({
                    session_id: sessionId,
                    turn_idx: turnIdx,
                    slug,
                    kind: 'mention',
                    ts,
                    prompt_excerpt_hash: lastPromptHash,
                });
            }
            listed = new Set();
        }
    }
    return records;
}

export function load_known_slugs(repo: string): Set<string> {
    const slugs = new Set<string>();
    const roots = [
        path.join(repo, '.augment', 'skills'),
        path.join(repo, '.claude', 'skills'),
        path.join(repo, 'dist/agent-src', 'skills'),
    ];
    for (const root of roots) {
        let entries: fs.Dirent[];
        try {
            entries = fs.readdirSync(root, { withFileTypes: true });
        } catch {
            continue;
        }
        for (const e of entries) {
            if (!e.isDirectory()) {
                continue;
            }
            const skillMd = path.join(root, e.name, 'SKILL.md');
            try {
                if (fs.statSync(skillMd).isFile()) {
                    slugs.add(e.name);
                }
            } catch {
                // no SKILL.md — skip.
            }
        }
    }
    return slugs;
}

export function dedup_key(rec: Record_): string {
    // Tuple identity (session_id, turn_idx, slug, kind). JSON-array string
    // key so the components cannot collide regardless of their contents.
    return JSON.stringify([rec.session_id, rec.turn_idx, rec.slug, rec.kind]);
}

// --- json.dumps(separators=(",", ":")) replica (compact, ensure_ascii) -------

function _jsonDumpsCompact(obj: Record<string, Json>): string {
    const enc = (value: Json): string => {
        if (value === null || value === undefined) return 'null';
        if (typeof value === 'boolean') return value ? 'true' : 'false';
        if (typeof value === 'number') return String(value);
        if (typeof value === 'string') return encStr(value);
        if (Array.isArray(value)) {
            return '[' + value.map(enc).join(',') + ']';
        }
        const o = value as { [k: string]: Json };
        const inner = Object.keys(o).map((k) => encStr(k) + ':' + enc(o[k] as Json));
        return '{' + inner.join(',') + '}';
    };
    const encStr = (s: string): string => {
        let out = '"';
        for (const ch of s) {
            const cp = ch.codePointAt(0) as number;
            if (ch === '"') out += '\\"';
            else if (ch === '\\') out += '\\\\';
            else if (ch === '\n') out += '\\n';
            else if (ch === '\r') out += '\\r';
            else if (ch === '\t') out += '\\t';
            else if (ch === '\b') out += '\\b';
            else if (ch === '\f') out += '\\f';
            else if (cp < 0x20) out += '\\u' + cp.toString(16).padStart(4, '0');
            else if (cp < 0x7f) out += ch;
            else if (cp > 0xffff) {
                const v = cp - 0x10000;
                const hi = 0xd800 + (v >> 10);
                const lo = 0xdc00 + (v & 0x3ff);
                out += '\\u' + hi.toString(16).padStart(4, '0');
                out += '\\u' + lo.toString(16).padStart(4, '0');
            } else {
                out += '\\u' + cp.toString(16).padStart(4, '0');
            }
        }
        return out + '"';
    };
    return enc(obj);
}

/**
 * Mirror Python `Path.relative_to(REPO)` — throws when `p` is not under REPO
 * (Python `relative_to` raises ValueError → exit 1; the traceback prose is
 * Python-version-dependent and not matched).
 */
function _relToRepoPosix(p: string): string {
    const rel = path.relative(REPO, p);
    if (rel === '..' || rel.startsWith('..' + path.sep) || path.isAbsolute(rel)) {
        throw new Error(`'${p}' is not in the subpath of '${REPO}'`);
    }
    return rel.split(path.sep).join('/');
}

interface ParsedArgs {
    project_slug: string | null;
    out: string;
    quiet: boolean;
}

function parse_args(argv: string[]): ParsedArgs {
    const out: ParsedArgs = { project_slug: null, out: OUT, quiet: false };
    for (let i = 0; i < argv.length; i += 1) {
        const a = argv[i] as string;
        const takeValue = (flag: string): string => {
            const eq = a.indexOf('=');
            if (eq !== -1) {
                return a.slice(eq + 1);
            }
            const next = argv[i + 1];
            if (next === undefined) {
                process.stderr.write(
                    `skill_usage_collect: error: argument ${flag}: expected one argument\n`,
                );
                process.exit(2);
            }
            i += 1;
            return next;
        };
        if (a === '--project-slug' || a.startsWith('--project-slug=')) {
            out.project_slug = takeValue('--project-slug');
        } else if (a === '--out' || a.startsWith('--out=')) {
            out.out = takeValue('--out');
        } else if (a === '--quiet') {
            out.quiet = true;
        } else if (a === '-h' || a === '--help') {
            process.stdout.write(
                'usage: skill_usage_collect [-h] [--project-slug PROJECT_SLUG] [--out OUT] [--quiet]\n',
            );
            process.exit(0);
        }
    }
    return out;
}

export function main(argv: string[] | null = null): number {
    const args = parse_args(argv ?? process.argv.slice(2));

    const slug = args.project_slug || project_slug(REPO);
    const files = session_files(slug);
    if (files.length === 0) {
        if (!args.quiet) {
            process.stderr.write(`no session files for slug ${slug}\n`);
        }
        return 0;
    }
    const known = load_known_slugs(REPO);
    const seen = new Set<string>();
    fs.mkdirSync(path.dirname(args.out), { recursive: true });
    if (fs.existsSync(args.out)) {
        const text = fs.readFileSync(args.out, 'utf-8');
        for (const line of text.split('\n')) {
            if (!line) {
                continue;
            }
            try {
                const rec = JSON.parse(line) as Partial<Record_>;
                if (
                    rec.session_id === undefined ||
                    rec.turn_idx === undefined ||
                    rec.slug === undefined ||
                    rec.kind === undefined
                ) {
                    continue;
                }
                seen.add(dedup_key(rec as Record_));
            } catch {
                continue;
            }
        }
    }
    let appended = 0;
    const fh = fs.openSync(args.out, 'a');
    try {
        for (const jsonl of files) {
            for (const rec of collect_session(jsonl, known)) {
                const k = dedup_key(rec);
                if (seen.has(k)) {
                    continue;
                }
                seen.add(k);
                fs.writeSync(
                    fh,
                    _jsonDumpsCompact(rec as unknown as Record<string, Json>) + '\n',
                );
                appended += 1;
            }
        }
    } finally {
        fs.closeSync(fh);
    }
    if (!args.quiet) {
        process.stdout.write(
            `✅  Wrote ${appended} new record(s) to ${_relToRepoPosix(path.resolve(args.out))} (${seen.size} total)\n`,
        );
    }
    return 0;
}

const _isCliEntry =
    process.argv[1] !== undefined &&
    import.meta.url === pathToFileURL(path.resolve(process.argv[1])).href;
if (_isCliEntry || process.argv[1] === _HERE) {
    process.exit(main());
}
